### ISSUE TITLE
[JENKINS-43336] Allow locking multiple stages in declarative pipeline

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/pipeline-model-definition-plugin#1043]
<p>It would be useful to be able to lock multiple stages as a single lock. For example, we usually have a stage to deploy to an environment and then another stage to run end-to-end tests on that environment, but there should be no other concurrent deployments until both stages have completed.</p>

<p>Something like this:</p>

<pre>
pipeline {
  stages {
    lock(resource: <span class="code-quote">'myResource'</span>, inversePrecedence: <span class="code-keyword">true</span>) {
      stage(<span class="code-quote">'Deploy'</span>) {
<span class="code-comment">// deploy to environment
</span>      }

      stage(<span class="code-quote">'E2E'</span>) {
<span class="code-comment">// run tests on the environment
</span>milestone 1
      }
    }
  }
}</pre>
<p>Technically both stages could just be merged into a single stage but to me that defeats the purpose of stages.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rochdev">rochdev</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336" target="_blank">Allow locking multiple stages in declarative pipeline</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: lockable-resources-plugin, pipeline-model-definition-plugin
<li><b>resolution</b>: Duplicate
<li><b>resolved</b>: 2017-12-18T17:52:35+00:00
<li><b>votes</b>: 19
<li><b>watchers</b>: 39
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>It would be useful to be able to lock multiple stages as a single lock. For example, we usually have a stage to deploy to an environment and then another stage to run end-to-end tests on that environment, but there should be no other concurrent deployments until both stages have completed.</p>

<p>Something like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">pipeline {
  stages {
    lock(resource: <span class="code-quote">'myResource'</span>, inversePrecedence: <span class="code-keyword">true</span>) {
      stage(<span class="code-quote">'Deploy'</span>) {
        <span class="code-comment">// deploy to environment
</span>      }

      stage(<span class="code-quote">'E2E'</span>) {
        <span class="code-comment">// run tests on the environment
</span>        milestone 1
      }
    }
  }
}</pre>
</div></div>
<p>Technically both stages could just be merged into a single stage but to me that defeats the purpose of stages.</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
- _Remote links associated with this issue:_


  - [Lockable Resources PR #​64](https://github.com/jenkinsci/lockable-resources-plugin/pull/64)
  - [tryLock and releaseLock](https://github.com/jenkinsci/lockable-resources-plugin/pull/179)

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dageissl:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=294577&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294577">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=294577&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294577">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>There may be some implementations in common with <a href="https://issues.jenkins-ci.org/browse/JENKINS-41334" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-41334</a> ? As there needs to be a more variable way of structuring stages.</p>
</pre>
</details>

</ul>
</details>

<p>There may be some implementations in common with <a href="https://issues.jenkins-ci.org/browse/JENKINS-41334" class="external-link" target="_blank" rel="nofollow noopener">https://issues.jenkins-ci.org/browse/JENKINS-41334</a> ? As there needs to be a more variable way of structuring stages.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>danturn:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=294579&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294579">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=294579&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294579">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi guys, </p>

<p>Is there any chance of this being implemented?</p>

<p>currently i have a pipeline where i create an apt repo on a linux agent, then pull from it on a different agent. I need those 2 stages to not be interrupted so ideally i would lock around both stages.</p>

<p>I'm not sure of a way to do this currently?</p>

<p> </p>

<p>dan</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>Hi guys, </p>

<p>Is there any chance of this being implemented?</p>

<p>currently i have a pipeline where i create an apt repo on a linux agent, then pull from it on a different agent. I need those 2 stages to not be interrupted so ideally i would lock around both stages.</p>

<p>I'm not sure of a way to do this currently?</p>

<p> </p>

<p>dan</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>danturn:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=294580&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294580">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=294580&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294580">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>for me actually an alternative would be to be able to run steps within a stage on different agents but i'm not sure that makes sense</p>
</pre>
</details>

</ul>
</details>

<p>for me actually an alternative would be to be able to run steps within a stage on different agents but i'm not sure that makes sense</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rochdev:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=294627&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294627">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=294627&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294627">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danturn" class="user-hover" rel="danturn">danturn</a> I didn't try it but if I remember correctly you can already run steps on different agents with something like this:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'My Stage'</span>) {
  steps {
    node(<span class="code-quote">'some label'</span>) {
      <span class="code-comment">// run step 1
</span>    }

    node(<span class="code-quote">'another label'</span>) {
      <span class="code-comment">// run step 2
</span>    }
  }
}</pre>
</div></div>
<p>You probably have to use `agent none` for this to work properly.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=danturn" class="user-hover" rel="danturn">danturn</a> I didn't try it but if I remember correctly you can already run steps on different agents with something like this:</p>

<pre>
stage(<span class="code-quote">'My Stage'</span>) {
  steps {
    node(<span class="code-quote">'some label'</span>) {
      <span class="code-comment">// run step 1
</span>    }

    node(<span class="code-quote">'another label'</span>) {
      <span class="code-comment">// run step 2
</span>    }
  }
}</pre>
<p>You probably have to use `agent none` for this to work properly.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>danturn:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=294633&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294633">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=294633&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294633">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>ooh ok, that would solve my problem <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>
 
i'll try it and update
 
thanks</p>
</pre>
</details>

</ul>
</details>

<p>ooh ok, that would solve my problem <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><br/>
 <br/>
i'll try it and update<br/>
 <br/>
thanks</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>danturn:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=294818&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294818">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=294818&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-294818">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>that works, thank you very much Roch!!</p>
</pre>
</details>

</ul>
</details>

<p>that works, thank you very much Roch!!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>matt_was:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=299434&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299434">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=299434&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299434">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Alternatively it would be great if we could do something like lock() somewhere and unlock() in another place in pipeline.</p>

<p>Example (similar to my own case):</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[...]
def selectedNode = utils.selectAvailableResource()

pipeline {
  agent none
    stages {
      stage(<span class="code-quote">'Checkout code'</span>) {
        agent { label selectedNode }
        steps {
          lock(<span class="code-quote">'aResource'</span>)
        }
      }
      stage(<span class="code-quote">'Build'</span>) {
        agent { label selectedNode }
        steps {
          echo <span class="code-quote">'Still locked'</span>
        }
      }
      stage(<span class="code-quote">'Integration tests'</span>) {
        agent { label selectedNode }
        steps {
          echo <span class="code-quote">'Still locked'</span>
        }
      }
      stage(<span class="code-quote">'Cleanup'</span>) {
        agent { label selectedNode }
        steps {
          unlock(<span class="code-quote">'aResource'</span>)
          echo <span class="code-quote">'Finally unlocked'</span>
        }
      }
    }
  }
}</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Alternatively it would be great if we could do something like lock() somewhere and unlock() in another place in pipeline.</p>

<p>Example (similar to my own case):</p>

<pre>
[...]
def selectedNode = utils.selectAvailableResource()

pipeline {
  agent none
    stages {
      stage(<span class="code-quote">'Checkout code'</span>) {
agent { label selectedNode }
steps {
  lock(<span class="code-quote">'aResource'</span>)
}
      }
      stage(<span class="code-quote">'Build'</span>) {
agent { label selectedNode }
steps {
  echo <span class="code-quote">'Still locked'</span>
}
      }
      stage(<span class="code-quote">'Integration tests'</span>) {
agent { label selectedNode }
steps {
  echo <span class="code-quote">'Still locked'</span>
}
      }
      stage(<span class="code-quote">'Cleanup'</span>) {
agent { label selectedNode }
steps {
  unlock(<span class="code-quote">'aResource'</span>)
  echo <span class="code-quote">'Finally unlocked'</span>
}
      }
    }
  }
}</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dageissl:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=299959&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299959">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=299959&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-299959">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matt_was" class="user-hover" rel="matt_was">matt_was</a> suggested would open up many exiting possibilities.</p>

<p>For example locking a resource prior to acquiring a node (best practice), but releasing it while retaining the build slot and workspace for further processing that does not need any locks.</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-comment">// acquire some kind of test resources
</span>def locks = lock(label: <span class="code-quote">'aResource'</span>, quantity: 1)
node {
    stage(<span class="code-quote">'build'</span>){
       sh <span class="code-quote">"mvn install -D${locks[0].name}"</span>
       unlock(locks)
    }
    <span class="code-comment">// continoue processing without having to leave the node
</span>    stage(<span class="code-quote">'analysis'</span>){
        sh <span class="code-quote">"mvn sonar:sonar"</span>
    }
}</pre>
</div></div>
<p>having that return value would be awesome too (to ask for the name of a resource and possibly more properties), but this is discussed in other tickets.</p>
</pre>
</details>

</ul>
</details>

<p>What <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=matt_was" class="user-hover" rel="matt_was">matt_was</a> suggested would open up many exiting possibilities.</p>

<p>For example locking a resource prior to acquiring a node (best practice), but releasing it while retaining the build slot and workspace for further processing that does not need any locks.</p>

<pre>
<span class="code-comment">// acquire some kind of test resources
</span>def locks = lock(label: <span class="code-quote">'aResource'</span>, quantity: 1)
node {
    stage(<span class="code-quote">'build'</span>){
       sh <span class="code-quote">"mvn install -D${locks[0].name}"</span>
       unlock(locks)
    }
    <span class="code-comment">// continoue processing without having to leave the node
</span>    stage(<span class="code-quote">'analysis'</span>){
sh <span class="code-quote">"mvn sonar:sonar"</span>
    }
}</pre>
<p>having that return value would be awesome too (to ask for the name of a resource and possibly more properties), but this is discussed in other tickets.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dolphy01:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=303497&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303497">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=303497&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303497">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is the single issue that is preventing us from moving to declarative pipeline. We currently use scripted Jenkinsfiles for exactly this purpose: to be able to define multiple stages which can lock a resource atomically.</p>

<p>In our case, we perform embedded hardware testing which requires the use of hardware resources to run testing. Because there are several stages in our pipeline which interact with the hardware (resetting, provisioning, testing, etc), the ability to lock across multiple stages is a necessity.</p>

<p>I do like the concept of being able to control "lock" versus "unlock". I think that would solve the problem sufficiently for us, and open up new possibilities.</p>
</pre>
</details>

</ul>
</details>

<p>This is the single issue that is preventing us from moving to declarative pipeline. We currently use scripted Jenkinsfiles for exactly this purpose: to be able to define multiple stages which can lock a resource atomically.</p>

<p>In our case, we perform embedded hardware testing which requires the use of hardware resources to run testing. Because there are several stages in our pipeline which interact with the hardware (resetting, provisioning, testing, etc), the ability to lock across multiple stages is a necessity.</p>

<p>I do like the concept of being able to control "lock" versus "unlock". I think that would solve the problem sufficiently for us, and open up new possibilities.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>scarlson:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=304848&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304848">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=304848&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-304848">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+1 ^^</p>

<p>This is also the only issue preventing us from using this in our CICD system.  We rely on locking to lock external resources so we can build a queue and they do not effect eachother on multiple pipelines (If there is another way that would suffice but nothing has worked as yet).  stage by stage does not suffice.</p>

<p> </p>

<p>The lock/unlock is also a highly desirable functionality.  We could make great use of it.</p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>+1 ^^</p>

<p>This is also the only issue preventing us from using this in our CICD system.  We rely on locking to lock external resources so we can build a queue and they do not effect eachother on multiple pipelines (If there is another way that would suffice but nothing has worked as yet).  stage by stage does not suffice.</p>

<p> </p>

<p>The lock/unlock is also a highly desirable functionality.  We could make great use of it.</p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=305341&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-305341">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=305341&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-305341">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Initial WIP PR in <tt>lockable-resources</tt> for adding <tt>getLock(...)</tt> and <tt>releaseLock()</tt> steps is up at <a href="https://github.com/jenkinsci/lockable-resources-plugin/pull/64" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/lockable-resources-plugin/pull/64</a>.</p>
</pre>
</details>

</ul>
</details>

<p>Initial WIP PR in <tt>lockable-resources</tt> for adding <tt>getLock(...)</tt> and <tt>releaseLock()</tt> steps is up at <a href="https://github.com/jenkinsci/lockable-resources-plugin/pull/64" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/lockable-resources-plugin/pull/64</a>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>anthonymastrean:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=308165&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308165">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=308165&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-308165">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Would we also be able to lock in the <tt>options</tt> block? To lock a resource for the entire build (and released implicitly)?</p>
</pre>
</details>

</ul>
</details>

<p>Would we also be able to lock in the <tt>options</tt> block? To lock a resource for the entire build (and released implicitly)?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>zaphinath:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=313280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-313280">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=313280&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-313280">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>+1 To resource request. We use traditional pipes that have several stages that need to share a lock.</p>
</pre>
</details>

</ul>
</details>

<p>+1 To resource request. We use traditional pipes that have several stages that need to share a lock.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=323114&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-323114">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=323114&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-323114">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The combination of <a href="https://issue-redirect.jenkins.io/issue/46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a> and <a href="https://issue-redirect.jenkins.io/issue/48380" title="Block-scoped &quot;wrapper&quot; options for individual stages" class="issue-link" data-issue-key="JENKINS-48380"><del>JENKINS-48380</del></a> will let you define a list of stages to run within a single top level (or parallel) stage, which will let you get what's desired here.</p>
</pre>
</details>

</ul>
</details>

<p>The combination of <a href="https://issue-redirect.jenkins.io/issue/46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a> and <a href="https://issue-redirect.jenkins.io/issue/48380" title="Block-scoped "wrapper" options for individual stages" class="issue-link" data-issue-key="JENKINS-48380"><del>JENKINS-48380</del></a> will let you define a list of stages to run within a single top level (or parallel) stage, which will let you get what's desired here.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>wortex17:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=328363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328363">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=328363&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328363">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> I'm not so sure about this, as this only works if we would be allowed to wrap the substages inside an "lock" declaration, which, as of now, would be impossible as lock is not a valid direct child of a stage (has to be nested inside steps).</p>

<p>I tried</p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Build Editor'</span>) {
    lock(<span class="code-quote">"myLock"</span>) {
        steps {
            <span class="code-comment">//...
</span>        }
    }
}
</pre>
</div></div>
<p>,</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Build Editor'</span>) {
    lock(<span class="code-quote">"myLock"</span>) {
        parallel {
            stage(<span class="code-quote">"ReallyBuild"</span>) {
                steps {
                    <span class="code-comment">//...
</span>                }
            }
        }
    }
}
</pre>
</div></div>
<p> as well as </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Build Editor'</span>) {
    parallel {
        lock(<span class="code-quote">"myLock"</span>) {
            stage(<span class="code-quote">"ReallyBuild"</span>) {
                steps {
                    <span class="code-comment">//...
</span>                }
            }
        }
    }
}
</pre>
</div></div>
<p> to no avail, as it fails with</p>
<blockquote><p>No "steps" or "parallel" to execute within stage</p></blockquote>
<p> or</p>
<blockquote><p>No stages specified ...</p></blockquote>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> I'm not so sure about this, as this only works if we would be allowed to wrap the substages inside an "lock" declaration, which, as of now, would be impossible as lock is not a valid direct child of a stage (has to be nested inside steps).</p>

<p>I tried</p>

<p> </p>

<pre>
stage(<span class="code-quote">'Build Editor'</span>) {
    lock(<span class="code-quote">"myLock"</span>) {
steps {
    <span class="code-comment">//...
</span>}
    }
}
</pre>
<p>,</p>

<pre>
stage(<span class="code-quote">'Build Editor'</span>) {
    lock(<span class="code-quote">"myLock"</span>) {
parallel {
    stage(<span class="code-quote">"ReallyBuild"</span>) {
steps {
    <span class="code-comment">//...
</span>}
    }
}
    }
}
</pre>
<p> as well as </p>

<pre>
stage(<span class="code-quote">'Build Editor'</span>) {
    parallel {
lock(<span class="code-quote">"myLock"</span>) {
    stage(<span class="code-quote">"ReallyBuild"</span>) {
steps {
    <span class="code-comment">//...
</span>}
    }
}
    }
}
</pre>
<p> to no avail, as it fails with</p>
<blockquote><p>No "steps" or "parallel" to execute within stage</p></blockquote>
<p> or</p>
<blockquote><p>No stages specified ...</p></blockquote>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=328364&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328364">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=328364&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-328364">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issue-redirect.jenkins.io/issue/46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a> will give you what you need on top of what's already available from <a href="https://issue-redirect.jenkins.io/issue/48380" title="Block-scoped &quot;wrapper&quot; options for individual stages" class="issue-link" data-issue-key="JENKINS-48380"><del>JENKINS-48380</del></a>, so that you could do something like:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Parent'</span>) {
  options {
    lock(<span class="code-quote">'myLock'</span>)
  }
  stages {
    stage(<span class="code-quote">'first child'</span>) {
      ...
    }
    stage(<span class="code-quote">'second child'</span>) {
      ...
    }
  }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p><a href="https://issue-redirect.jenkins.io/issue/46809" title="Allow sequential stages inside parallel in Declarative syntax" class="issue-link" data-issue-key="JENKINS-46809"><del>JENKINS-46809</del></a> will give you what you need on top of what's already available from <a href="https://issue-redirect.jenkins.io/issue/48380" title="Block-scoped "wrapper" options for individual stages" class="issue-link" data-issue-key="JENKINS-48380"><del>JENKINS-48380</del></a>, so that you could do something like:</p>


<pre>
stage(<span class="code-quote">'Parent'</span>) {
  options {
    lock(<span class="code-quote">'myLock'</span>)
  }
  stages {
    stage(<span class="code-quote">'first child'</span>) {
      ...
    }
    stage(<span class="code-quote">'second child'</span>) {
      ...
    }
  }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rrueth:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=360708&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-360708">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=360708&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-360708">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a>, I'm trying to lock multiple stages based on the Jenkins node on which the stages will execute. It seems like the only way to do this is using the `options` block that you specified:</p>


<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Parent'</span>) {
  options {
    lock(<span class="code-quote">"${env.NODE_NAME}"</span>)
  }
  stages {
    stage(<span class="code-quote">'first child'</span>) {
      ...
    }
    stage(<span class="code-quote">'second child'</span>) {
      ...
    }
  }
}</pre>
</div></div>
<p>Unfortunately, the resource this attempts to lock is `null`:</p>
<blockquote><p>Trying to acquire lock on <span class="error">&#91;null&#93;</span></p></blockquote>
<p> </p>

<p>It appears that this won't work because the stage's `options` directive occurs before entering the agent. The <a href="https://jenkins.io/doc/book/pipeline/syntax/#stage-options" class="external-link" target="_blank" rel="nofollow noopener">declarative pipeline syntax  documentation</a> says: </p>
<blockquote><p>Inside a <tt>stage</tt>, the steps in the <tt>options</tt> directive are invoked before entering the <tt>agent</tt> or checking any <tt>when</tt> conditions.</p></blockquote>
<p> 
Is there any way to lock multiple stages for a single env.NODE_NAME?  Ideally, other subsequent stages could still run in parallel with other builds (i.e. so limiting executors to 1 or disabling concurrent builds are not options).
 </p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a>, I'm trying to lock multiple stages based on the Jenkins node on which the stages will execute. It seems like the only way to do this is using the `options` block that you specified:</p>



<pre>
stage(<span class="code-quote">'Parent'</span>) {
  options {
    lock(<span class="code-quote">"${env.NODE_NAME}"</span>)
  }
  stages {
    stage(<span class="code-quote">'first child'</span>) {
      ...
    }
    stage(<span class="code-quote">'second child'</span>) {
      ...
    }
  }
}</pre>
<p>Unfortunately, the resource this attempts to lock is `null`:</p>
<blockquote><p>Trying to acquire lock on <span class="error">&#91;null&#93;</span></p></blockquote>
<p> </p>

<p>It appears that this won't work because the stage's `options` directive occurs before entering the agent. The <a href="https://jenkins.io/doc/book/pipeline/syntax/#stage-options" class="external-link" target="_blank" rel="nofollow noopener">declarative pipeline syntax  documentation</a> says: </p>
<blockquote><p>Inside a <tt>stage</tt>, the steps in the <tt>options</tt> directive are invoked before entering the <tt>agent</tt> or checking any <tt>when</tt> conditions.</p></blockquote>
<p> <br/>
Is there any way to lock multiple stages for a single env.NODE_NAME?  Ideally, other subsequent stages could still run in parallel with other builds (i.e. so limiting executors to 1 or disabling concurrent builds are not options).<br/>
 </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nikolodus:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=362486&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362486">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=362486&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-362486">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Just wanted to +1 on <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rrueth" class="user-hover" rel="rrueth">rrueth</a>'s comment regarding locking multiple stages with lock named after $NODE_NAME. Pretty much in the same situation, and I'd like to know if this can be supported - either by enabling options block to execute after agent has been assigned for a given stage, or through some other means if this is not an option.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> do you know if this could be possible?</p>
</pre>
</details>

</ul>
</details>

<p>Just wanted to +1 on <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rrueth" class="user-hover" rel="rrueth">rrueth</a>'s comment regarding locking multiple stages with lock named after $NODE_NAME. Pretty much in the same situation, and I'd like to know if this can be supported - either by enabling options block to execute after agent has been assigned for a given stage, or through some other means if this is not an option.</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> do you know if this could be possible?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>philip_abraham:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=364097&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364097">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=364097&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364097">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><tt>env</tt> is not available in <tt>options</tt> I think, but you could try getting the node name from the API:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage(<span class="code-quote">'Parent'</span>) {
   options {
     lock(<span class="code-quote">"${currentBuild.getRawBuild().getExecutor().getOwner().getDisplayName()}"</span>)
   }
   stages {
      stage(<span class="code-quote">'first child'</span>) {
         ...
      }
      stage(<span class="code-quote">'second child'</span>) {
         ...
      }
   }
}</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p><tt>env</tt> is not available in <tt>options</tt> I think, but you could try getting the node name from the API:</p>

<pre>
stage(<span class="code-quote">'Parent'</span>) {
   options {
     lock(<span class="code-quote">"${currentBuild.getRawBuild().getExecutor().getOwner().getDisplayName()}"</span>)
   }
   stages {
      stage(<span class="code-quote">'first child'</span>) {
 ...
      }
      stage(<span class="code-quote">'second child'</span>) {
 ...
      }
   }
}</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>voz:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=364548&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364548">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=364548&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-364548">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi, just wanted to add that we have the same problem where we need to lock multiple stages at once for atomicity (environment build and integration tests go together). Also, we need to parametrise the lock with the environment against which the build is done: it is ok to have parallel builds against different environments, but not against the same one.</p>

<p>As it looks, currently the multistage lock is available only in the options section which does not support environment parameters. Hence, it looks that we can not achieve what we need.</p>
</pre>
</details>

</ul>
</details>

<p>Hi, just wanted to add that we have the same problem where we need to lock multiple stages at once for atomicity (environment build and integration tests go together). Also, we need to parametrise the lock with the environment against which the build is done: it is ok to have parallel builds against different environments, but not against the same one.</p>

<p>As it looks, currently the multistage lock is available only in the options section which does not support environment parameters. Hence, it looks that we can not achieve what we need.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/bitwiseman.png" title="bitwiseman's avatar" /> bitwiseman:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/43336?focusedId=378429&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-378429">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-43336?focusedId=378429&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-378429">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Bulk closing resolved issues. </p>
</pre>
</details>

</ul>
</details>

<p>Bulk closing resolved issues. </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-48380%22">JENKINS-48380</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `duplicates` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-46809%22">JENKINS-46809</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/pipeline-model-definition-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-45138%22">JENKINS-45138</a>]</i>
